### PR TITLE
84 `hash`, `uuid` custom functions | 83 `datetime` custom functions

### PR DESCRIPTION
**Pull request by vd2org**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# IDEs
```
Comment (spicy-sauce) [inline .gitignore]: let's use this instead:
```
# intellij configs
```

(same as done in the other project)

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,7 @@
...
+
```
Comment (spicy-sauce) [inline .gitignore]: remove empty line

File: datayoga-py/README.md [hunk 3]
```diff
@@ -124,14 +124,18 @@ DataYoga supports both SQL and [JMESPath](https://jmespath.org/) expressions. JM
...
+| `time_delta_days`    | Returns the number of days between given `dt` and now(positive) or the number of days that have passed from now(negative).       | Input: `{"dt": '2021-10-06T18:56:16.701670+00:00'}` <br /> Expression: `time_delta_days(dt)` <br /> Output: `365`                                 | If `dt` is string ISO datetime(2011-11-04T00:05:23+04:00) is assumed. If `dt` is number unix timestamp is assumed.
+| `time_delta_seconds` | Returns the number of seconds between given `dt` and now(positive) or the number of seconds that have passed from now(negative). | Input: `{"dt": '2021-10-06T18:56:16.701670+00:00'}` <br /> Expression: `time_delta_days(dt)` <br /> Output: `31557600`                            | If `dt` is string ISO datetime(2011-11-04T00:05:23+04:00) is assumed. If `dt` is number unix timestamp is assumed.
```
Comment (spicy-sauce) [inline datayoga-py/README.md]: place those new custom functions ordered alphabetically in the table.

Comment (spicy-sauce) [inline datayoga-py/README.md]: remove period from the description of these new 4 custom functions as in the other functions.
It's not common to end a description in a table (behaves as a title) with a period, see example:
https://support.microsoft.com/en-us/office/excel-functions-alphabetical-b3944572-255d-4efb-bb96-c6d90033e188

File: datayoga-py/README.md [hunk 4]
```diff
@@ -124,14 +124,18 @@ DataYoga supports both SQL and [JMESPath](https://jmespath.org/) expressions. JM
...
+| `split`              | Splits a string into a list of strings after breaking the given string by the specified delimiter (comma by default)             | Input: `{"departments": "finance,hr,r&d"}` <br /> Expression: `split(departments)` <br /> Output: `['finance', 'hr', 'r&d']`                      | Default delimiter is comma - a different delimiter can be passed to the function as the second argument, for example: `split(departments, ';')` |
+| `upper`              | Converts all lowercase characters in a string into uppercase characters                                                          | Input: `{"fname": "john"}` <br /> Expression: `upper(fname)` <br /> Output: `JOHN`                                                                |
+| `uuid`               | Generates random uuid4 and returns it as string in standard format.                                                              | Input: None <br /> Expression: `uuid()` <br /> Output: `3264b35c-ff5d-44a8-8bc7-9be409dac2b7`                                                     |
```
Comment (spicy-sauce) [inline datayoga-py/README.md]: returns it as string --> returns it as a string

Comment (spicy-sauce) [inline datayoga-py/README.md]: uuid4 --> UUID4

Comment (spicy-sauce) [inline datayoga-py/README.md]: Generates random -->  Generates a random

File: datayoga-py/README.md [hunk 5]
```diff
@@ -124,14 +124,18 @@ DataYoga supports both SQL and [JMESPath](https://jmespath.org/) expressions. JM
...
+| `hash`               | Calculates hash using given `hash_name` hash function and returns hexadecimal representation.                                    | Input: `{"some_str": "some_value"}` <br /> Expression: `hash(some_str, &#96;sha1&#96;)` <br /> Output: `8c818171573b03feeae08b0b4ffeb6999e3afc05` | Supported hash functions: sha256, md5, sha384, sha3_384, blake2b, sha1, sha512, sha3_224, sha224, sha3_256, sha3_512, blake2s
```
Comment (spicy-sauce) [inline datayoga-py/README.md]: Calculates hash using given `hash_name` hash function and returns hexadecimal representation --> Calculates a hash using given the `hash_name` hash function and returns its hexadecimal representation

File: datayoga-py/README.md [hunk 6]
```diff
@@ -124,14 +124,18 @@ DataYoga supports both SQL and [JMESPath](https://jmespath.org/) expressions. JM
...
+| `time_delta_days`    | Returns the number of days between given `dt` and now(positive) or the number of days that have passed from now(negative).       | Input: `{"dt": '2021-10-06T18:56:16.701670+00:00'}` <br /> Expression: `time_delta_days(dt)` <br /> Output: `365`                                 | If `dt` is string ISO datetime(2011-11-04T00:05:23+04:00) is assumed. If `dt` is number unix timestamp is assumed.
```
Comment (spicy-sauce) [inline datayoga-py/README.md]: If `dt` is number unix timestamp is assumed --> If `dt` is a number, Unix timestamp is assumed

Comment (spicy-sauce) [inline datayoga-py/README.md]: If `dt` is a string,  ISO datetime (2011-11-04T00:05:23+04:00, for example) is assumed

File: datayoga-py/README.md (hunk 3 above)
Comment (spicy-sauce) [inline datayoga-py/README.md]: If dt is number unix timestamp is assumed --> If dt is a number, Unix timestamp is assumed

Comment (spicy-sauce) [inline datayoga-py/README.md]: If dt is a string, ISO datetime (2011-11-04T00:05:23+04:00, for example) is assumed

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 7]
```diff
@@ -1,5 +1,11 @@
+
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: redundant line

File: datayoga-py/README.md (hunk 6 above)
Comment (spicy-sauce) [inline datayoga-py/README.md]: add an example to number as well.

File: datayoga-py/README.md (hunk 3 above)
Comment (spicy-sauce) [inline datayoga-py/README.md]: add an example to number as well.

File: datayoga-py/README.md (hunk 6 above)
Comment (spicy-sauce) [inline datayoga-py/README.md]: What are the number of days that have passed from now?

File: datayoga-py/README.md (hunk 3 above)
Comment (spicy-sauce) [inline datayoga-py/README.md]: What are the number of days that have passed from now?

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 8]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
     @functions.signature({"types": ["string", "null"], "variadic": True})
     def _func_split(self, element, delimiter=","):
         return str(element).split(delimiter) if element is not None else None
+
+    @functions.signature()
+    def _func_uuid(self):
+        """Generates random uuid4 and returns it as string in standard format."""
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: Generates a random UUID4 and returns it as a string in standard format

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 9]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+
+        return str(uuid4())
+
+    @functions.signature({"types": ["number", "string", "boolean", "array", "object", "null"], "variadic": True})
+    def _func_hash(self, obj, hash_name="sha1"):
+        """\
+        Calculates hash using given `hash_name` hash function and returns hexadecimal representation.
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: Calculates a hash using given the hash_name hash function and returns its hexadecimal representation

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 10]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+
+    @functions.signature()
+    def _func_uuid(self):
+        """Generates random uuid4 and returns it as string in standard format."""
+
+        return str(uuid4())
+
+    @functions.signature({"types": ["number", "string", "boolean", "array", "object", "null"], "variadic": True})
+    def _func_hash(self, obj, hash_name="sha1"):
+        """\
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py:58]: \ is redundant

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 11]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+        Calculates hash using given `hash_name` hash function and returns hexadecimal representation.
+
+        The following algorithms were tested:
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: I'd say: 
Supported algorithms:

File: datayoga-py/README.md (hunk 5 above)
Comment (spicy-sauce) [inline datayoga-py/README.md]: mention that `sha1` is the default if not specified.

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 12]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+        h.update(prepare())
+        return h.hexdigest()
+
+    @functions.signature({"types": ["string", "number"]})
+    def _func_time_delta_days(self, dt):
+        """\
+        Returns the number of days between given `dt` and now(positive)
+        or the number of days that have passed from now(negative).
+
+        If `dt` is string ISO datetime(2011-11-04T00:05:23+04:00) is assumed.
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: If `dt` is a string, ISO datetime(2011-11-04T00:05:23+04:00) is assumed.

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 13]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+        If `dt` is number unix timestamp is assumed.
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: If `dt` is a number,  Unix timestamp is assumed.

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 14]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+        """
+
+        dt = datetime.fromisoformat(dt) if isinstance(dt, str) else datetime.fromtimestamp(dt)
+
+        now = dt.now(dt.tzinfo)
+
+        delta = now - dt
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: I'd avoid empty line between each line

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 15]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+
+        return delta.days
+
+    @functions.signature({"types": ["string", "number"]})
+    def _func_time_delta_seconds(self, dt):
+        """\
+        Returns the number of seconds between given `dt` and now(positive)
+        or the number of seconds that have passed from now(negative).
+
+        If `dt` is string ISO datetime(2011-11-04T00:05:23+04:00) is assumed.
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: If `dt` is a string, ISO datetime(2011-11-04T00:05:23+04:00) is assumed.

File: datayoga-py/datayoga/jmespath_custom_functions.py [hunk 16]
```diff
@@ -41,3 +47,83 @@ def _func_mid(self, element, offset, amount):
...
+        If `dt` is number unix timestamp is assumed.
```
Comment (spicy-sauce) [inline datayoga-py/datayoga/jmespath_custom_functions.py]: If `dt` is a number, Unix timestamp is assumed.

**Review by spicy-sauce — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (vd2org) [reply .gitignore]: ✅

File: .gitignore (hunk 2 above)
Comment (vd2org) [reply .gitignore]: ✅

File: datayoga-py/README.md (hunk 3 above)
Comment (vd2org) [reply datayoga-py/README.md]: ✅

Comment (vd2org) [reply datayoga-py/README.md]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 15 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 16 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 13 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 12 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/README.md (hunk 5 above)
Comment (vd2org) [reply datayoga-py/README.md]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 11 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 10 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py:58]: This slash will remove the first empty line in the doctoring.
I prefer to format like this for multiline docstring to keep identical indentation of the text.

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 9 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 8 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 7 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/README.md (hunk 3 above)
Comment (vd2org) [reply datayoga-py/README.md]: ✅

Comment (vd2org) [reply datayoga-py/README.md]: ✅

File: datayoga-py/README.md (hunk 6 above)
Comment (vd2org) [reply datayoga-py/README.md]: ✅

Comment (vd2org) [reply datayoga-py/README.md]: ✅

File: datayoga-py/README.md (hunk 5 above)
Comment (vd2org) [reply datayoga-py/README.md]: ✅

File: datayoga-py/README.md (hunk 4 above)
Comment (vd2org) [reply datayoga-py/README.md]: ✅

Comment (vd2org) [reply datayoga-py/README.md]: ✅

Comment (vd2org) [reply datayoga-py/README.md]: ✅

File: datayoga-py/datayoga/jmespath_custom_functions.py (hunk 14 above)
Comment (vd2org) [reply datayoga-py/datayoga/jmespath_custom_functions.py]: ✅

File: datayoga-py/README.md (hunk 6 above)
Comment (vd2org) [reply datayoga-py/README.md]: If the `dt` value in the past the number will be positive otherwise negative.

Comment (vd2org) [reply datayoga-py/README.md]: ✅

File: datayoga-py/README.md (hunk 3 above)
Comment (vd2org) [reply datayoga-py/README.md]: ✅

**Review by spicy-sauce — APPROVED** [review]

**Pull request merged**